akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -219,18 +237,22 @@ object GraphStageLogic {
               if (!_watchedBy.compareAndSet(watchedBy, watchedBy - watcher))
                 remWatcher(watchee, watcher) // try again
           } else if (!watcheeSelf && watcherSelf) {
-            unwatch(watchee)
+            log.warning("externally triggered unwatch from {} to {} is illegal on StageActorRef", watcher, watchee)
           } else {
             log.error("BUG: illegal Unwatch(%s,%s) for %s".format(watchee, watcher, this))
           }
       }
     }
 
-    def watch(actorRef: ActorRef): Unit =
+    def watch(actorRef: ActorRef): Unit = {
+      watching += actorRef
       actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Watch(actorRef.asInstanceOf[InternalActorRef], this))
+    }
 
-    def unwatch(actorRef: ActorRef): Unit =
+    def unwatch(actorRef: ActorRef): Unit = {
+      watching -= actorRef
```

Comment:
I guess the assumption is those are called from within the Stage, so it's synchronous and safe.
// spelling it out for my understanding as well